File: Java/695_Max_Area_of_Island.java
class Solution {
    int [][] grid;
    boolean [][] seen;
    public int maxAreaOfIsland(int[][] grid) {
        this.grid = grid;
        seen = new boolean [grid.length][grid[0].length];
        
        int count = 0;
        
        for(int i=0; i<grid.length; i++){
            for(int j=0; j<grid[0].length; j++){
                count = Math.max(count, dfs(i,j));
            }
        }
        return count;
    }
    
    public int dfs(int r, int c){
        if(r<0 || c<0 || r>=grid.length || c>=grid[0].length || seen[r][c] || grid[r][c] == 0 ) return 0;
        
        seen[r][c] = true;
        
        return( 1 + dfs(r+1,c) + dfs(r-1,c) + dfs(r,c+1)+ dfs(r, c-1) );
    }
}

File: Python/695. Max Area Of Island.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        def explore(grid, r, c, visited):
            rowInbound = 0 <= r <len(grid)
            colInbound = 0 <= c < len(grid[0])
            
            if not(rowInbound) or not(colInbound):return 0
            if grid[r][c] == 0: return 0
            elem = str(r)+','+str(c)
            if visited.__contains__(elem): return 0
            visited.add(elem)
          
            
            return 1 + explore(grid,r-1,c,visited) + explore(grid,r+1,c,visited)+explore(grid,r,c+1,visited) + explore(grid,r,c-1,visited)
        
        visited = set()
        max_c = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                max_c = max(max_c,explore(grid,i,j,visited))
        return max_c
        